src/Auth/Credentials/CredentialsTrait.php

```diff
@@ -0,0 +1,274 @@
+    {
+        $result = $this->fetchAuthToken();
+        if (isset($result['access_token'])) {
+            return ['Authorization' => 'Bearer ' . $result['access_token']];
+        }
+
+        return [];
+    }
+
+    /**
+     * Implements CredentialsInterface#fetchAuthToken.
+     *
+     * Fetches the auth tokens and caches it based on the default cache
+     * configuration.
+     *
+     * @throws \Exception
+     *
+     * @return array The auth token
+     *
+     * Access tokens have the following keys:
+     *   - access_token (string)
+     *   - expires_in (int)
+     *   - token_type (string)
+     * ID tokens have the following keys:
+     *   - id_token (string)
+     */
+    public function fetchAuthToken(): array
+    {
+        $cacheKey = $this->getCacheKey();
+        if ($cachedToken = $this->getCachedToken($cacheKey)) {
+            return $cachedToken;
+        }
+
+        $token = $this->fetchAuthTokenNoCache();
+
+        $this->setCachedToken($cacheKey, $token);
+
+        return $token;
+    }
+
+    /**
+     * @param array           $options
+     * @param ClientInterface $options.httpClient
+     */
+    private function setHttpClientFromOptions(array $options): void
+    {
+        if (empty($options['httpClient'])) {
+            $options['httpClient'] = ClientFactory::build();
+        }
+        if (!$options['httpClient'] instanceof ClientInterface) {
+            throw new \RuntimeException(sprintf(
+                'Invalid option "httpClient": must be an instance of %s',
+                ClientInterface::class
+            ));
+        }
+        $this->httpClient = $options['httpClient'];
+    }
+
+    /**
+     * @param array                  $options
+     * @param CacheItemPoolInterface $options.cache
+     * @param int                    $options.cacheLifetime
+     * @param strring                $options.cachePrefix
+     */
+    private function setCacheFromOptions(array $options): void
+    {
+        if (!empty($options['cache'])) {
+            if (!$options['cache'] instanceof CacheItemPoolInterface) {
+                throw new \RuntimeException(sprintf(
+                    'Invalid option "cache": must be an instance of %s',
+                    CacheItemPoolInterface::class
+                ));
+            }
+            $this->cache = $options['cache'];
+        } else {
+            $this->cache = new MemoryCacheItemPool();
+        }
+        if (array_key_exists('cacheLifetime', $options)) {
+            $this->cacheLifetime = (int) $options['cacheLifetime'];
+        }
+        if (array_key_exists('cachePrefix', $options)) {
+            $this->cachePrefix = (string) $options['cachePrefix'];
+        }
+    }
+
+    private function getCacheKey(): string
+    {
+        throw new \LogicException(
+            'getCacheKey must be implemented in the Credentials class'
+        );
+    }
+
+    private function fetchAuthTokenNoCache(): array
+    {
+        throw new \LogicException(
```

Comment:
[Traits can have abstract methods](https://www.php.net/manual/en/language.oop5.traits.php#language.oop5.traits.abstract), would that make more sense here?

Reply:
That's a good point. It might make sense to make them abstract, but I was thinking that since both these functions are optional (credentials don't _have_ to support caching) that it makes more sense to throw the exception at runtime. In the case of `AnonymousCredentials`, there's no way for those exceptions to be thrown because they'd never be called.